jetty-core/jetty-http2/jetty-http2-common/src/main/java/org/eclipse/jetty/http2/internal/HTTP2Flusher.java

```diff
@@ -248,6 +257,15 @@ protected Action process() throws Throwable
                     entry.failed(failure);
                     pending.remove();
                 }
+                catch (HpackException.SessionException failure)
+                {
+                    if (LOG.isDebugEnabled())
+                        LOG.debug("Failure generating {}", entry, failure);
+                    onSessionFailure(failure);
+                    // The method above will try to send
+                    // a GOAWAY, so we will iterate again.
+                    return Action.IDLE;
```

Comment:
I've walked through the code of `onSessionFailure` and `fail(Throwable)` and it is not obvious what is sending the go-away, so it is not clear what will call iterated again? More importantly, it is not clear if that will always be done?

Is there some way of making that clearer here?

Reply:
`onSessionFailure()` attempts to call `session.close()`, which may send a GOAWAY (if not already closed, etc.).

If the GOAWAY is sent, then the flusher is iterated, and returning IDLE will result in a re-iteration.
Otherwise the flusher is already terminated, and we can return IDLE, as it will be torn down soon after anyway.